File: obstacle_avoidance.py
# Importing all required ros2 in-built packages
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import LaserScan
from geometry_msgs.msg import Twist
from rclpy.qos import QoSProfile
from rclpy.qos import qos_profile_sensor_data

# Defining intial distance to obstacle as 1
distToObstacle = 1

class ObstacleAvoidance(Node):
    def __init__(self):
        super().__init__('obstacle_avoidance')

        qos = QoSProfile(depth=10)

        # Creating a publisher to publish the robot's velocity and a subscriber to subscribe to laser scan messages
        self.sub = self.create_subscription(LaserScan, '/scan', self.obstacle_callback, qos_profile=qos_profile_sensor_data)
        self.pub = self.create_publisher(Twist, '/cmd_vel', 10)


    def obstacle_callback(self, msg):
        self.get_logger().info('The distance to obstacle is %s ' % msg.ranges[300])

        move = Twist()

        # Defining conditions for the robot to change direction when it encounters an obstacle
        if msg.ranges[300] > distToObstacle:
            move.linear.x = 0.5
            move.angular.z = 0.0

        if msg.ranges[300] <= distToObstacle:
            move.linear.x = 0.5
            move.angular.z = 1.5 
      
        self.pub.publish(move)

def main(args=None):

    rclpy.init(args=args)

    obstacle_avoidance = ObstacleAvoidance()

    rclpy.spin(obstacle_avoidance)

    obstacle_avoidance.destroy_node()

    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: obstacle_avoidance.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "sensor_msgs/msg/laser_scan.hpp"
#include "rclcpp/qos.hpp"

constexpr double distToObstacle = 1.0;

class ObstacleAvoidance : public rclcpp::Node
{
public:
    ObstacleAvoidance()
        : Node("obstacle_avoidance")
    {
        auto qos = rclcpp::QoS(rclcpp::KeepLast(10));

        subscription_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
            "/scan",
            qos,
            std::bind(&ObstacleAvoidance::obstacle_callback, this, std::placeholders::_1));

        publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("/cmd_vel", 10);
    }

private:
    void obstacle_callback(const sensor_msgs::msg::LaserScan::SharedPtr msg)
    {
        RCLCPP_INFO(this->get_logger(), "The distance to obstacle is %f ", msg->ranges[300]);

        auto move = geometry_msgs::msg::Twist();

        if (msg->ranges[300] > distToObstacle)
        {
            move.linear.x = 0.5;
            move.angular.z = 0.0;
        }

        if (msg->ranges[300] <= distToObstacle)
        {
            move.linear.x = 0.5;
            move.angular.z = 3.5;
        }

        publisher_->publish(move);
    }

    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr subscription_;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);

    auto obstacle_avoidance = std::make_shared<ObstacleAvoidance>();

    rclcpp::spin(obstacle_avoidance);

    rclcpp::shutdown();

    return 0;
}

